tests/test_signals.py

```diff
@@ -0,0 +1,255 @@
+    @app.signal("foo.bar.baz")
+    def sync_signal(*_):
+        ...
+
+    @app.signal("foo.bar.baz")
+    async def async_signal(*_):
+        ...
+
+    assert len(app.signal_router.routes) == 1
+
+
+@pytest.mark.parametrize(
+    "signal",
+    (
+        "<foo>.bar.bax",
+        "foo.<bar>.baz",
+        "foo",
+        "foo.bar",
+        "foo.bar.baz.qux",
+    ),
+)
+def test_invalid_signal(app, signal):
+    with pytest.raises(InvalidSignal, match=f"Invalid signal event: {signal}"):
+
+        @app.signal(signal)
+        def handler():
+            ...
+
+
+@pytest.mark.asyncio
+async def test_dispatch_signal_triggers_multiple_handlers(app):
+    counter = 0
+
+    @app.signal("foo.bar.baz")
+    def sync_signal(*_):
+        nonlocal counter
+
+        counter += 1
+
+    @app.signal("foo.bar.baz")
+    async def async_signal(*_):
+        nonlocal counter
+
+        counter += 1
+
+    app.signal_router.finalize()
+
+    await app.dispatch("foo.bar.baz")
+    assert counter == 2
+
+    await app.dispatch("foo", "bar", "baz")
+    assert counter == 4
+
+
+@pytest.mark.asyncio
+async def test_dispatch_signal_triggers_triggers_event(app):
+    counter = 0
+
+    @app.signal("foo.bar.baz")
+    def sync_signal(*args):
+        nonlocal app
+        nonlocal counter
+        signal, *_ = app.signal_router.get("foo.bar.baz")
+        counter += signal.ctx.event.is_set()
+
+    app.signal_router.finalize()
+
+    await app.dispatch("foo.bar.baz")
+    signal, *_ = app.signal_router.get("foo.bar.baz")
+
+    assert counter == 1
+
+
+@pytest.mark.asyncio
+async def test_dispatch_signal_triggers_dynamic_route(app):
+    counter = 0
+
+    @app.signal("foo.bar.<baz:int>")
+    def sync_signal(baz):
+        nonlocal counter
+
+        counter += baz
+
+    app.signal_router.finalize()
+
+    await app.dispatch("foo.bar.9")
+    assert counter == 9
+
+
+@pytest.mark.asyncio
+async def test_dispatch_signal_triggers_with_requirements(app):
+    counter = 0
+
+    @app.signal("foo.bar.baz", requirements={"one": "two"})
+    def sync_signal(*_):
+        nonlocal counter
+        counter += 1
+
+    app.signal_router.finalize()
+
+    await app.dispatch("foo.bar.baz")
+    assert counter == 0
+    await app.dispatch("foo.bar.baz", where={"one": "two"})
+    assert counter == 1
+
+
+@pytest.mark.asyncio
+async def test_dispatch_signal_triggers_with_context(app):
+    counter = 0
+
+    @app.signal("foo.bar.baz")
+    def sync_signal(amount):
+        nonlocal counter
+        counter += amount
+
+    app.signal_router.finalize()
+
+    await app.dispatch("foo.bar.baz", context={"amount": 9})
+    assert counter == 9
+
+
+@pytest.mark.asyncio
+async def test_dispatch_signal_triggers_with_context_fail(app):
+    counter = 0
+
+    @app.signal("foo.bar.baz")
+    def sync_signal(amount):
+        nonlocal counter
+        counter += amount
+
+    app.signal_router.finalize()
+
+    task = await app.dispatch("foo.bar.baz", {"amount": 9})
```

Comment:
I think this should still be an acceptable way to dispatch a signal. We can avoid the `*fields` in the `dispatch` function and allow this way to sending the context instead? This is also more user friendly.

Reply:
If we get rid of the exploded strings, then yes. We can support this.

Reply:
I think that maybe force kwarg here:

```python
    def dispatch(
        self,
        event: str,
        *,
        where: Optional[Dict[str, str]] = None,
        context: Optional[Dict[str, Any]] = None,
    ):
```

In this example:

```python
     task = await app.dispatch("foo.bar.baz", {"amount": 9}) 
```

It is ambiguous what was intended and easy to mixup.

Reply:
I see what you did there. I am good with what you have done now. :)